### ISSUE TITLE
error when make run 

### ISSUE THREAD

**Issue by michealylj1** [p4lang/tutorials#253]
Hi All,

I build one P4 environment and just run the command "make run" and meet the error message as:

ljyang@Tang:~/P4/tutorials/exercises/basic$ make run 
mkdir -p build pcaps logs
p4c-bm2-ss --p4v 16 --p4runtime-files build/basic.p4.p4info.txt -o build/basic.json basic.p4
terminate called after throwing an instance of 'std::system_error'
  what():  Unknown error -1
../../utils/Makefile:35: recipe for target 'build/basic.json' failed
make: *** [build/basic.json] Aborted (core dumped)
ljyang@Tang:~/P4/tutorials/exercises/basic$ 

Please tell me how to fix it?

Thanks,
Michael

**Comment by jnfoster** [comment]
Are you on the [tutorial VM](https://drive.google.com/uc?id=1lYF4NgFkYoRqtskdGTMxy3sXUV0jkMxo&export=download)?

**Comment by michealylj1** [comment]
No, I run it my host environment. 
I just rebuild the P4 environment by the script file "user-bootstrap.sh", but I still meet another issue as:

ljyang@Tang:~/P4/tutorials/exercises/basic$ make run
mkdir -p build pcaps logs
sudo python ../../utils/run_exercise.py -t topology.json -b simple_switch_grpc
Traceback (most recent call last):
  File "../../utils/run_exercise.py", line 33, in <module>
    import p4runtime_lib.simple_controller
  File "/home/ljyang/P4/tutorials/utils/p4runtime_lib/simple_controller.py", line 22, in <module>
    import bmv2
  File "/home/ljyang/P4/tutorials/utils/p4runtime_lib/bmv2.py", line 15, in <module>
    from switch import SwitchConnection
  File "/home/ljyang/P4/tutorials/utils/p4runtime_lib/switch.py", line 20, in <module>
    from p4.v1 import p4runtime_pb2
  File "/usr/local/lib/python2.7/dist-packages/p4/v1/p4runtime_pb2.py", line 18, in <module>
    from google.rpc import status_pb2 as google_dot_rpc_dot_status__pb2
  File "/usr/local/lib/python2.7/dist-packages/google/rpc/status_pb2.py", line 92, in <module>
    from grpc.beta import implementations as beta_implementations
  File "/usr/local/lib/python2.7/dist-packages/grpc/beta/implementations.py", line 22, in <module>
    from grpc import _auth
  File "/usr/local/lib/python2.7/dist-packages/grpc/_auth.py", line 40, in <module>
    class GoogleCallCredentials(grpc.AuthMetadataPlugin):
AttributeError: 'module' object has no attribute 'AuthMetadataPlugin'
../../utils/Makefile:27: recipe for target 'run' failed
make: *** [run] Error 1

**Comment by jnfoster** [comment]
Hmm. These scripts were designed to work on the VM but YMMV if you are starting from a different, arbitrary Unix or Unix-like environment. 

Perhaps you could provide more details about your environment:
* What Linux distribution are you using? What version of that distribution?
* What is the GitHub hash of the VM install scripts you used?
* What versions of GRPC and the Python grpcio package do you have installed?

**Comment by michealylj1** [comment]
The Linux distribution is  Ubuntu 16.04.6 LTS,  the kernel version as: 4.4.0-143-generic #169-Ubuntu SMP Thu Feb 7 07:56:38 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

It's same as the script file "user-bootstrap.sh" as:

BMV2_COMMIT="884e01b531c6fd078cc2438a40258ecae011a65b"  # Apr 24, 2019
PI_COMMIT="19de33e83bae7b737a3f8a1c9507c6e84173d96f"    # Apr 24, 2019
P4C_COMMIT="61409c890c58d14ec7d6790f263eb44f393e542a"   # Apr 24, 2019
PROTOBUF_COMMIT="v3.2.0"
GRPC_COMMIT="v1.3.2"

**Comment by jnfoster** [comment]
The VM has a newer version of `grpcio`

```
p4@p4:~$ pip show grpcio
---
Metadata-Version: 2.1
Name: grpcio
Version: 1.20.1
Summary: HTTP/2-based RPC framework
Home-page: https://grpc.io
Author: The gRPC Authors
Author-email: grpc-io@googlegroups.com
Installer: pip
License: Apache License 2.0
Location: /usr/local/lib/python2.7/dist-packages
Requires: six, futures, enum34
...
```
Can you try to upgrade to that version and see if it resolves your issue?

**Comment by michealylj1** [comment]
I just upgrade grpcio to 1.20.1 and the issue has been solved, but I meet another issue as:

ljyang@Tang:~/P4/tutorials/exercises/basic$ make run
mkdir -p build pcaps logs
p4c-bm2-ss --p4v 16 --p4runtime-files build/basic.p4.p4info.txt -o build/basic.json basic.p4
basic.p4(81): [--Wwarn=unused] warning: Unused action parameter dstAddr
    action ipv4_forward(macAddr_t dstAddr, egressSpec_t port) {
                                  ^^^^^^^
basic.p4(81): [--Wwarn=unused] warning: Unused action parameter port
    action ipv4_forward(macAddr_t dstAddr, egressSpec_t port) {
                                                        ^^^^
sudo python ../../utils/run_exercise.py -t topology.json -b simple_switch_grpc
Reading topology file.
Building mininet topology.
Switch port mapping:
s1:  1:h1       2:s2    3:s3
s2:  1:h2       2:s1    3:s3
s3:  1:h3       2:s1    3:s2
P4 switch s1 did not start correctly.
../../utils/Makefile:27: recipe for target 'run' failed
make: *** [run] Error 1

**Comment by jnfoster** [comment]
I suspect something is still wrong with your GRPC installation. For some reason, the `simple_switch_grpc` did not start up correctly.

Can you do three things:

1. Modify the utils/mininet/p4_mininet.py file to print the expression

> `' '.join(args)`

just before this line, which prints the error message

https://github.com/p4lang/tutorials/blob/01fc3780760118ac227ba18724b9dba644d15c9a/utils/p4_mininet.py#L148

2. Clean up mininet 

> `mn -c`

3. Re-run the command

> ` sudo python ../../utils/run_exercise.py -t topology.json -b simple_switch_grpc`

and print the output.

**Comment by michealylj1** [comment]
ljyang@Tang:~/P4/tutorials/exercises/basic$ sudo python ../../utils/run_exercise.py -t topology.json -b simple_switch_grpc
Reading topology file.
Building mininet topology.
Switch port mapping:
s1:  1:h1       2:s2    3:s3
s2:  1:h2       2:s1    3:s3
s3:  1:h3       2:s1    3:s2
Traceback (most recent call last):
  File "../../utils/run_exercise.py", line 408, in <module>
    exercise.run_exercise()
  File "../../utils/run_exercise.py", line 201, in run_exercise
    self.create_network()
  File "../../utils/run_exercise.py", line 266, in create_network
    controller = None)
  File "build/bdist.linux-x86_64/egg/mininet/net.py", line 174, in __init__
  File "build/bdist.linux-x86_64/egg/mininet/net.py", line 502, in build
  File "build/bdist.linux-x86_64/egg/mininet/net.py", line 489, in buildFromTopo
  File "build/bdist.linux-x86_64/egg/mininet/net.py", line 400, in addLink
  File "build/bdist.linux-x86_64/egg/mininet/link.py", line 557, in __init__
  File "build/bdist.linux-x86_64/egg/mininet/link.py", line 451, in __init__
  File "build/bdist.linux-x86_64/egg/mininet/link.py", line 496, in makeIntfPair
  File "build/bdist.linux-x86_64/egg/mininet/util.py", line 239, in makeIntfPair
Exception: Error creating interface pair (s1-eth2,s2-eth2): RTNETLINK answers: File exists
 
ljyang@Tang:~/P4/tutorials/exercises/basic$

**Comment by jnfoster** [comment]
It looks like you didn't call `print`. Change the line you added to

> `print ' '.join(args)`

and try again?

**Comment by michealylj1** [comment]
It shout be change the file " ../../utils/p4runtime_switch.py".

ljyang@Tang:~/P4/tutorials/exercises/basic$ git diff ../../utils/p4runtime_switch.py
diff --git a/utils/p4runtime_switch.py b/utils/p4runtime_switch.py
index a3b8bcf..50c3558 100644
--- a/utils/p4runtime_switch.py
+++ b/utils/p4runtime_switch.py
@@ -131,6 +131,7 @@ class P4RuntimeSwitch(P4Switch):
             pid = int(f.read())
         debug("P4 switch {} PID is {}.\n".format(self.name, pid))
         if not self.check_switch_started(pid):
+           print ' '.join(args)
             error("P4 switch {} did not start correctly.\n".format(self.name))
             exit(1)
         info("P4 switch {} has been started.\n".format(self.name))
ljyang@Tang:~/P4/tutorials/exercises/basic$ 
ljyang@Tang:~/P4/tutorials/exercises/basic$ sudo mn -c
*** Removing excess controllers/ofprotocols/ofdatapaths/pings/noxes
killall controller ofprotocol ofdatapath ping nox_corelt-nox_core ovs-openflowd ovs-controllerovs-testcontroller udpbwtest mnexec ivs ryu-manager 2> /dev/null
killall -9 controller ofprotocol ofdatapath ping nox_corelt-nox_core ovs-openflowd ovs-controllerovs-testcontroller udpbwtest mnexec ivs ryu-manager 2> /dev/null
pkill -9 -f "sudo mnexec"
*** Removing junk from /tmp
rm -f /tmp/vconn* /tmp/vlogs* /tmp/*.out /tmp/*.log
*** Removing old X11 tunnels
*** Removing excess kernel datapaths
ps ax | egrep -o 'dp[0-9]+' | sed 's/dp/nl:/'
***  Removing OVS datapaths
ovs-vsctl --timeout=1 list-br
ovs-vsctl --timeout=1 list-br
*** Removing all links of the pattern foo-ethX
ip link show | egrep -o '([-_.[:alnum:]]+-eth[[:digit:]]+)'
( ip link del s2-eth2;ip link del s1-eth2;ip link del s1-eth2;ip link del s2-eth2;ip link del s3-eth2;ip link del s1-eth3;ip link del s1-eth3;ip link del s3-eth2;ip link del s3-eth3;ip link del s2-eth3;ip link del s2-eth3;ip link del s3-eth3 ) 2> /dev/null
ip link show
*** Killing stale mininet node processes
pkill -9 -f mininet:
*** Shutting down stale tunnels
pkill -9 -f Tunnel=Ethernet
pkill -9 -f .ssh/mn
rm -f ~/.ssh/mn/*
*** Cleanup complete.
ljyang@Tang:~/P4/tutorials/exercises/basic$ sudo python ../../utils/run_exercise.py -t topology.json -b simple_switch_grpc
Reading topology file.
Building mininet topology.
Switch port mapping:
s1:  1:h1       2:s2    3:s3
s2:  1:h2       2:s1    3:s3
s3:  1:h3       2:s1    3:s2
simple_switch_grpc -i 1@s1-eth1 -i 2@s1-eth2 -i 3@s1-eth3 --pcap /home/ljyang/P4/tutorials/exercises/basic/pcaps --nanolog ipc:///tmp/bm-0-log.ipc --device-id 0 --no-p4 --log-console --thrift-port 9090 -- --grpc-server-addr 0.0.0.0:50051
P4 switch s1 did not start correctly.
ljyang@Tang:~/P4/tutorials/exercises/basic$

**Comment by jnfoster** [comment]
And what happens if you run that command directly?

**Comment by michealylj1** [comment]
ljyang@Tang:~/P4/tutorials/exercises/basic$ make clean
sudo mn -c
[sudo] password for ljyang: 
*** Removing excess controllers/ofprotocols/ofdatapaths/pings/noxes
killall controller ofprotocol ofdatapath ping nox_corelt-nox_core ovs-openflowd ovs-controllerovs-testcontroller udpbwtest mnexec ivs ryu-manager 2> /dev/null
killall -9 controller ofprotocol ofdatapath ping nox_corelt-nox_core ovs-openflowd ovs-controllerovs-testcontroller udpbwtest mnexec ivs ryu-manager 2> /dev/null
pkill -9 -f "sudo mnexec"
*** Removing junk from /tmp
rm -f /tmp/vconn* /tmp/vlogs* /tmp/*.out /tmp/*.log
*** Removing old X11 tunnels
*** Removing excess kernel datapaths
ps ax | egrep -o 'dp[0-9]+' | sed 's/dp/nl:/'
***  Removing OVS datapaths
ovs-vsctl --timeout=1 list-br
ovs-vsctl --timeout=1 list-br
*** Removing all links of the pattern foo-ethX
ip link show | egrep -o '([-_.[:alnum:]]+-eth[[:digit:]]+)'
( ip link del s2-eth2;ip link del s1-eth2;ip link del s1-eth2;ip link del s2-eth2;ip link del s3-eth2;ip link del s1-eth3;ip link del s1-eth3;ip link del s3-eth2;ip link del s3-eth3;ip link del s2-eth3;ip link del s2-eth3;ip link del s3-eth3 ) 2> /dev/null
ip link show
*** Killing stale mininet node processes
pkill -9 -f mininet:
*** Shutting down stale tunnels
pkill -9 -f Tunnel=Ethernet
pkill -9 -f .ssh/mn
rm -f ~/.ssh/mn/*
*** Cleanup complete.
rm -f *.pcap
rm -rf build pcaps logs
ljyang@Tang:~/P4/tutorials/exercises/basic$ 
ljyang@Tang:~/P4/tutorials/exercises/basic$ 
ljyang@Tang:~/P4/tutorials/exercises/basic$ make run
mkdir -p build pcaps logs
p4c-bm2-ss --p4v 16 --p4runtime-files build/basic.p4.p4info.txt -o build/basic.json basic.p4
basic.p4(81): [--Wwarn=unused] warning: Unused action parameter dstAddr
    action ipv4_forward(macAddr_t dstAddr, egressSpec_t port) {
                                  ^^^^^^^
basic.p4(81): [--Wwarn=unused] warning: Unused action parameter port
    action ipv4_forward(macAddr_t dstAddr, egressSpec_t port) {
                                                        ^^^^
sudo python ../../utils/run_exercise.py -t topology.json -b simple_switch_grpc
Reading topology file.
Building mininet topology.
Switch port mapping:
s1:  1:h1       2:s2    3:s3
s2:  1:h2       2:s1    3:s3
s3:  1:h3       2:s1    3:s2
simple_switch_grpc -i 1@s1-eth1 -i 2@s1-eth2 -i 3@s1-eth3 --pcap /home/ljyang/P4/tutorials/exercises/basic/pcaps --nanolog ipc:///tmp/bm-0-log.ipc --device-id 0 --no-p4 --log-console --thrift-port 9090 -- --grpc-server-addr 0.0.0.0:50051
P4 switch s1 did not start correctly.
../../utils/Makefile:27: recipe for target 'run' failed
make: *** [run] Error 1
ljyang@Tang:~/P4/tutorials/exercises/basic$

**Comment by jnfoster** [comment]
Sorry, I meant

```
simple_switch_grpc -i 1@s1-eth1 -i 2@s1-eth2 -i 3@s1-eth3 --pcap /home/ljyang/P4/tutorials/exercises/basic/pcaps --nanolog ipc:///tmp/bm-0-log.ipc --device-id 0 --no-p4 --log-console --thrift-port 9090 -- --grpc-server-addr 0.0.0.0:50051
```

**Comment by michealylj1** [comment]
ljyang@Tang:~/P4/tutorials/exercises/basic$ simple_switch_grpc -i 1@s1-eth1 -i 2@s1-eth2 -i 3@s1-eth3 --pcap /home/ljyang/P4/tutorials/exercises/basic/pcaps --nanolog ipc:///tmp/bm-0-log.ipc --device-id 0 --no-p4 --log-console --thrift-port 9090 -- --grpc-server-addr 0.0.0.0:50051
[libprotobuf FATAL google/protobuf/stubs/common.cc:67] This program requires version 3.6.1 of the Protocol Buffer runtime library, but the installed version is 3.2.0.  Please update your library.  If you compiled the program yourself, make sure that your headers are from the same version of Protocol Buffers as your link-time library.  (Version verification failed in "google/protobuf/any.pb.cc".)
terminate called after throwing an instance of 'google::protobuf::FatalException'
  what():  This program requires version 3.6.1 of the Protocol Buffer runtime library, but the installed version is 3.2.0.  Please update your library.  If you compiled the program yourself, make sure that your headers are from the same version of Protocol Buffers as your link-time library.  (Version verification failed in "google/protobuf/any.pb.cc".)
Aborted (core dumped)
ljyang@Tang:~/P4/tutorials/exercises/basic$ 

I know the reason of the issue is the wrong version of protocol buffer. But there are another version 3.6.1 for other project. Please tell me how to specify the libray path when I run the command "make run".

Thanks,
Michael

**Comment by michealylj1** [comment]
OK，Thanks...
